# Homework 6

## Problem 10

We are given 10 input units, 1 output unit and 36 units in the hidden layer (a unit also includes the bias node). We have to find an architecture for the hidden layer that **maximizes** the number of weight edges.

### Solution

The minimum number of weight edges is achieved for the architecture $d$ with 2 hidden layers, where each layer consists of a bias node and a regular node:

$d = [9, 21, 13, 1]$

For this architecture the number of weight edges is 510.

In [1]:
def get_edges(d):
    '''
    - Takes an architecture d.
    - Returns the number of edges in the neural network.
    '''
    total_edges = 0
    L = len(d) - 1
    for i in range(L):
        total_edges += (d[i] + 1) * d[i+1]
    return total_edges


print("total number of edges: ", get_edges([9, 21, 13, 1]))

total number of edges:  510


![neural_network_max_edges](figures/hw6_p10_neural_network_max_edges.png)

## Brute force approach via compositions

In order to find this architecture we will examine what the number of maximum weight edges is for a given number of hidden layers. 

- We have at least one hidden layer.

- Each hidden layer requires at least (1 bias + 1 regular node) = 2 hidden units. So having 36 hidden units at disposal means we can have at most $36/2 = 18$ hidden layers.

- We determine the maximum number of edges we can achieve with 1 hidden layer, with 2 hidden layers, ..., 18 hidden layers.

## 1 hidden layer

For 1 hidden layer there is only one possible architecture, namely all 36 units in the hidden layer. This means we have 1 bias node and 35 regular nodes  in the hidden layer, and the architecture is:

$d = [9, 35, 1]$

In [2]:
print("total number of edges = ", get_edges([9, 35, 1]))

total number of edges =  386


### Result

For 1 hidden layer the maximum number of edges is 386.

## 2 hidden layers

### 2a) Compositions

For 2 hidden layers we have 2 bias nodes. This leaves us with 36 - 2 = 34 regular nodes to distribute where each hidden layer contains at least 1 regular node.

- $(1, 33)$
- $(2, 32)$
- ...
- $(33, 1)$

This is related to the number of [compositions](https://en.wikipedia.org/wiki/Composition_(combinatorics) which is the number of ways we can represent an integer $n$ as a sum of nonnegative integers.


This yields the following possible architectures:

- $d = [9, 1, 33, 1]$
- $d = [9, 2, 32, 1]$
- ...
- $d = [9, 33, 1, 1]$

The number of compositions of $n$ into $k$ parts is given by ${n-1} \choose {k-1}$. In our case we have $n = 34$ and $k=2$, and the number of compositions is ${{n-1} \choose {k-1}} = {{33} \choose {1}} = 33$. We can use this number to check if we have considered all possible architectures.


In [3]:
from math import factorial


def n_choose_k(n, k):
    '''
    - Returns the binomial coefficient (n choose k),
      see https://en.wikipedia.org/wiki/Binomial_coefficient#Factorial_formula
    '''
    return factorial(n) // (factorial(k) * factorial(n-k))
    
    
def num_compositions(n, k):
    '''
    - Returns the number of compositions of n into k parts,
      see https://en.wikipedia.org/wiki/Composition_(combinatorics)
    - The number is ((n-1) choose (k-1))
    '''
    return n_choose_k(n-1, k-1)

### 2b) Generate all possible architectures for 2 hidden layers

Let's write a Python program that generates all possible architectures for 2 hidden layers.

In [4]:
def get_architectures_2_hidden_layers():
    my_list = []
    
    k = 2                   # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    for d1 in range(1, n):
        d2 = n-d1
        d = [9, d1, d2, 1]
        my_list.append(d)
        
    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_2 = get_architectures_2_hidden_layers()
print(*architectures_2, sep='\n')

[9, 1, 33, 1]
[9, 2, 32, 1]
[9, 3, 31, 1]
[9, 4, 30, 1]
[9, 5, 29, 1]
[9, 6, 28, 1]
[9, 7, 27, 1]
[9, 8, 26, 1]
[9, 9, 25, 1]
[9, 10, 24, 1]
[9, 11, 23, 1]
[9, 12, 22, 1]
[9, 13, 21, 1]
[9, 14, 20, 1]
[9, 15, 19, 1]
[9, 16, 18, 1]
[9, 17, 17, 1]
[9, 18, 16, 1]
[9, 19, 15, 1]
[9, 20, 14, 1]
[9, 21, 13, 1]
[9, 22, 12, 1]
[9, 23, 11, 1]
[9, 24, 10, 1]
[9, 25, 9, 1]
[9, 26, 8, 1]
[9, 27, 7, 1]
[9, 28, 6, 1]
[9, 29, 5, 1]
[9, 30, 4, 1]
[9, 31, 3, 1]
[9, 32, 2, 1]
[9, 33, 1, 1]



Let's furthermore determine the number of edges for each architecture.

In [5]:
for d in architectures_2:
    print("d = {0} has {1} edges".format(d, get_edges(d)))

d = [9, 1, 33, 1] has 110 edges
d = [9, 2, 32, 1] has 149 edges
d = [9, 3, 31, 1] has 186 edges
d = [9, 4, 30, 1] has 221 edges
d = [9, 5, 29, 1] has 254 edges
d = [9, 6, 28, 1] has 285 edges
d = [9, 7, 27, 1] has 314 edges
d = [9, 8, 26, 1] has 341 edges
d = [9, 9, 25, 1] has 366 edges
d = [9, 10, 24, 1] has 389 edges
d = [9, 11, 23, 1] has 410 edges
d = [9, 12, 22, 1] has 429 edges
d = [9, 13, 21, 1] has 446 edges
d = [9, 14, 20, 1] has 461 edges
d = [9, 15, 19, 1] has 474 edges
d = [9, 16, 18, 1] has 485 edges
d = [9, 17, 17, 1] has 494 edges
d = [9, 18, 16, 1] has 501 edges
d = [9, 19, 15, 1] has 506 edges
d = [9, 20, 14, 1] has 509 edges
d = [9, 21, 13, 1] has 510 edges
d = [9, 22, 12, 1] has 509 edges
d = [9, 23, 11, 1] has 506 edges
d = [9, 24, 10, 1] has 501 edges
d = [9, 25, 9, 1] has 494 edges
d = [9, 26, 8, 1] has 485 edges
d = [9, 27, 7, 1] has 474 edges
d = [9, 28, 6, 1] has 461 edges
d = [9, 29, 5, 1] has 446 edges
d = [9, 30, 4, 1] has 429 edges
d = [9, 31, 3, 1] has 410

### 2c) Determine maximum architecture for 2 hidden layers


Let's determine the architecture that achieves the maximum number of edges.

In [6]:
def get_max_architecture(list_archs):
    '''
    - Takes a list of architectures.
    - Returns the architecture with the maximum number of edges and the
      and the corresponding number of edges.
    '''
    max_edges = -1
    max_arch = None
    
    for d in list_archs:
        num_edges = get_edges(d)
        if max_edges < num_edges:
            max_edges = num_edges
            max_arch = d
    return max_arch, max_edges

#----------------------------

max_arch, max_edges = get_max_architecture(architectures_2)
print("For 2 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 2 hidden layers the architecture [9, 21, 13, 1] yields a maximum of 510 edges.


### Result

For 2 hidden layers the architecture [9, 21, 13, 1] yields a maximum of 510 edges.

## 3 hidden layers

### 3a) Compositions

For 3 hidden layers we have 3 bias nodes. This leaves us with 36 - 3 = 33 regular nodes to distribute where each hidden layer contains at least 1 regular node.

- $(1, 1, 31)$
- $(1, 2, 30)$
- $(1, 3, 29)$
- ...
- $(1, 31, 1)$
- $(2, 1, 30)$
- $(2, 2, 29)$
- $(2, 3, 28)$
- ...
- $(30, 2, 1)$
- $(31, 1, 1)$


This yields the following possible architectures:

- $d = [9, 1, 1, 31, 1]$
- $d = [9, 1, 2, 30, 1]$
- $d = [9, 1, 3, 29, 1]$
- ...
- $d = [9, 1, 31, 1, 1]$
- $d = [9, 2, 1, 30, 1]$
- $d = [9, 2, 2, 29, 1]$
- $d = [9, 2, 3, 28, 1]$
- ...
- $d = [9, 30, 2, 1, 1]$
- $d = [9, 31, 1, 1, 1]$

### 3b) Generate all possible architectures for 3 hidden layers

Let's write a Python program that generates all possible architectures for 3 hidden layers.

In [7]:
def get_architectures_3_hidden_layers():
    my_list = []
    
    k = 3                   # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            d3 = n-d1-d2
            d = [9, d1, d2, d3, 1]
            my_list.append(d)
        
    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_3 = get_architectures_3_hidden_layers()
print(*architectures_3[:3], sep='\n')
print("...")
print(*architectures_3[-3:], sep='\n')

[9, 1, 1, 31, 1]
[9, 1, 2, 30, 1]
[9, 1, 3, 29, 1]
...
[9, 30, 1, 2, 1]
[9, 30, 2, 1, 1]
[9, 31, 1, 1, 1]


### 3c) Determine maximum architecture for 3 hidden layers

In [8]:
max_arch, max_edges = get_max_architecture(architectures_3)
print("For 3 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 3 hidden layers the architecture [9, 20, 12, 1, 1] yields a maximum of 467 edges.


### Result

For 3 hidden layers the architecture [9, 20, 12, 1, 1] yields a maximum of 467 edges.

## 4 hidden layers

### 4.1 Generate all possible architectures for 4 hidden layers

In [9]:
def get_architectures_4_hidden_layers():
    my_list = []
    
    k = 4                   # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2) nodes are at our disposal
                d4 = n-d1-d2-d3
                d = [9, d1, d2, d3, d4, 1]
                my_list.append(d)
        
    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_4 = get_architectures_4_hidden_layers()
print(*architectures_3[:4], sep='\n')
print("...")
print(*architectures_3[-4:], sep='\n')

[9, 1, 1, 31, 1]
[9, 1, 2, 30, 1]
[9, 1, 3, 29, 1]
[9, 1, 4, 28, 1]
...
[9, 29, 3, 1, 1]
[9, 30, 1, 2, 1]
[9, 30, 2, 1, 1]
[9, 31, 1, 1, 1]


### 4.2 Determine maximum architecture for 4 hidden layers

In [10]:
max_arch, max_edges = get_max_architecture(architectures_4)
print("For 4 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 4 hidden layers the architecture [9, 19, 11, 1, 1, 1] yields a maximum of 426 edges.


### Result

For 4 hidden layers the architecture [9, 19, 11, 1, 1, 1] yields a maximum of 426 edges.

## 5 hidden layers

### 5.1 Generate all possible architectures for 5 hidden layers

In [11]:
def get_architectures_5_hidden_layers():
    my_list = []
    
    k = 5                   # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):

                    d5 = n-d1-d2-d3-d4
                    d = [9, d1, d2, d3, d4, d5, 1]
                    my_list.append(d)
        
    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_5 = get_architectures_5_hidden_layers()
print(*architectures_5[:4], sep='\n')
print("...")
print(*architectures_5[-4:], sep='\n')

[9, 1, 1, 1, 1, 27, 1]
[9, 1, 1, 1, 2, 26, 1]
[9, 1, 1, 1, 3, 25, 1]
[9, 1, 1, 1, 4, 24, 1]
...
[9, 26, 1, 1, 2, 1, 1]
[9, 26, 1, 2, 1, 1, 1]
[9, 26, 2, 1, 1, 1, 1]
[9, 27, 1, 1, 1, 1, 1]


### 5.2 Determine maximum architecture for 5 hidden layers

In [12]:
max_arch, max_edges = get_max_architecture(architectures_5)
print("For 5 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 5 hidden layers the architecture [9, 18, 10, 1, 1, 1, 1] yields a maximum of 387 edges.


### Result

For 5 hidden layers the architecture [9, 18, 10, 1, 1, 1, 1] yields a maximum of 387 edges.

## 6 hidden layers

### 6.1 Generate all possible architectures for 6 hidden layers¶

In [13]:
def get_architectures_6_hidden_layers():
    my_list = []
    
    k = 6                   # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):

                            d6 = n-d1-d2-d3-d4-d5
                            d = [9, d1, d2, d3, d4, d5, d6, 1]
                            my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_6 = get_architectures_6_hidden_layers()
print(*architectures_6[:4], sep='\n')
print("...")
print(*architectures_6[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 25, 1]
[9, 1, 1, 1, 1, 2, 24, 1]
[9, 1, 1, 1, 1, 3, 23, 1]
[9, 1, 1, 1, 1, 4, 22, 1]
...
[9, 24, 1, 1, 2, 1, 1, 1]
[9, 24, 1, 2, 1, 1, 1, 1]
[9, 24, 2, 1, 1, 1, 1, 1]
[9, 25, 1, 1, 1, 1, 1, 1]


### 6.2 Determine maximum architecture for 6 hidden layers

In [14]:
max_arch, max_edges = get_max_architecture(architectures_6)
print("For 6 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 6 hidden layers the architecture [9, 17, 9, 1, 1, 1, 1, 1] yields a maximum of 350 edges.


### Result

For 6 hidden layers the architecture [9, 17, 9, 1, 1, 1, 1, 1] yields a maximum of 350 edges.

## 7 hidden layers

### 7.1 Generate all possible architectures for 7 hidden layers¶

In [15]:
def get_architectures_7_hidden_layers():
    my_list = []
    
    k = 7                   # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):
                            # for d6 only (n-d1-d2-d3-d4-d5) nodes are at our disposal
                            for d6 in range(1, n-d1-d2-d3-d4-d5):

                                d7 = n-d1-d2-d3-d4-d5-d6
                                d = [9, d1, d2, d3, d4, d5, d6, d7, 1]
                                my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_7 = get_architectures_7_hidden_layers()
print(*architectures_7[:4], sep='\n')
print("...")
print(*architectures_7[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 1, 23, 1]
[9, 1, 1, 1, 1, 1, 2, 22, 1]
[9, 1, 1, 1, 1, 1, 3, 21, 1]
[9, 1, 1, 1, 1, 1, 4, 20, 1]
...
[9, 22, 1, 1, 2, 1, 1, 1, 1]
[9, 22, 1, 2, 1, 1, 1, 1, 1]
[9, 22, 2, 1, 1, 1, 1, 1, 1]
[9, 23, 1, 1, 1, 1, 1, 1, 1]


### 7.2 Determine maximum architecture for 7 hidden layers

In [16]:
max_arch, max_edges = get_max_architecture(architectures_7)
print("For 7 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 7 hidden layers the architecture [9, 16, 8, 1, 1, 1, 1, 1, 1] yields a maximum of 315 edges.


## 8 hidden layers

### 8.1 Generate all possible architectures for 8 hidden layers¶

In [17]:
def get_architectures_8_hidden_layers():
    my_list = []
    
    k = 8                   # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):
                            # for d6 only (n-d1-d2-d3-d4-d5) nodes are at our disposal
                            for d6 in range(1, n-d1-d2-d3-d4-d5):
                                # for d7 only (n-d1-d2-d3-d4-d5-d6) nodes are at our disposal
                                for d7 in range(1, n-d1-d2-d3-d4-d5-d6):
                                
                                    d8 = n-d1-d2-d3-d4-d5-d6-d7
                                    d = [9, d1, d2, d3, d4, d5, d6, d7, d8, 1]
                                    my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_8 = get_architectures_8_hidden_layers()
print(*architectures_8[:4], sep='\n')
print("...")
print(*architectures_8[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 1, 1, 21, 1]
[9, 1, 1, 1, 1, 1, 1, 2, 20, 1]
[9, 1, 1, 1, 1, 1, 1, 3, 19, 1]
[9, 1, 1, 1, 1, 1, 1, 4, 18, 1]
...
[9, 20, 1, 1, 2, 1, 1, 1, 1, 1]
[9, 20, 1, 2, 1, 1, 1, 1, 1, 1]
[9, 20, 2, 1, 1, 1, 1, 1, 1, 1]
[9, 21, 1, 1, 1, 1, 1, 1, 1, 1]


### 8.2 Determine maximum architecture for 8 hidden layers

In [18]:
max_arch, max_edges = get_max_architecture(architectures_8)
print("For 8 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 8 hidden layers the architecture [9, 15, 7, 1, 1, 1, 1, 1, 1, 1] yields a maximum of 282 edges.


### Result

For 8 hidden layers the architecture [9, 15, 7, 1, 1, 1, 1, 1, 1, 1] yields a maximum of 282 edges.

## 9 hidden layers

### 9.1 Generate all possible architectures for 9 hidden layers¶

In [19]:
def get_architectures_9_hidden_layers():
    my_list = []
    
    k = 9                   # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):
                            # for d6 only (n-d1-d2-d3-d4-d5) nodes are at our disposal
                            for d6 in range(1, n-d1-d2-d3-d4-d5):
                                # for d7 only (n-d1-d2-d3-d4-d5-d6) nodes are at our disposal
                                for d7 in range(1, n-d1-d2-d3-d4-d5-d6):
                                    # for d8 only (n-d1-d2-d3-d4-d5-d6-d7) nodes are at our disposal
                                    for d8 in range(1, n-d1-d2-d3-d4-d5-d6-d7):
                                    
                                        d9 = n-d1-d2-d3-d4-d5-d6-d7-d8
                                        d = [9, d1, d2, d3, d4, d5, d6, d7, d8, d9, 1]
                                        my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_9 = get_architectures_9_hidden_layers()
print(*architectures_9[:4], sep='\n')
print("...")
print(*architectures_9[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 1, 1, 1, 19, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 2, 18, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 3, 17, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 4, 16, 1]
...
[9, 18, 1, 1, 2, 1, 1, 1, 1, 1, 1]
[9, 18, 1, 2, 1, 1, 1, 1, 1, 1, 1]
[9, 18, 2, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 19, 1, 1, 1, 1, 1, 1, 1, 1, 1]


### 9.2 Determine maximum architecture for 9 hidden layers

In [20]:
max_arch, max_edges = get_max_architecture(architectures_9)
print("For 9 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 9 hidden layers the architecture [9, 14, 6, 1, 1, 1, 1, 1, 1, 1, 1] yields a maximum of 251 edges.


## 10 hidden layers

### 10.1 Generate all possible architectures for 10 hidden layers¶

In [21]:
def get_architectures_10_hidden_layers():
    my_list = []
    
    k = 10                  # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):
                            # for d6 only (n-d1-d2-d3-d4-d5) nodes are at our disposal
                            for d6 in range(1, n-d1-d2-d3-d4-d5):
                                # for d7 only (n-d1-d2-d3-d4-d5-d6) nodes are at our disposal
                                for d7 in range(1, n-d1-d2-d3-d4-d5-d6):
                                    # for d8 only (n-d1-d2-d3-d4-d5-d6-d7) nodes are at our disposal
                                    for d8 in range(1, n-d1-d2-d3-d4-d5-d6-d7):
                                        # for d9 only (n-d1-d2-d3-d4-d5-d6-d7-d8) nodes are at our disposal
                                        for d9 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8):

                                            d10 = n-d1-d2-d3-d4-d5-d6-d7-d8-d9
                                            d = [9, d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, 1]
                                            my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_10 = get_architectures_10_hidden_layers()
print(*architectures_10[:4], sep='\n')
print("...")
print(*architectures_10[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 17, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 2, 16, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 3, 15, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 4, 14, 1]
...
[9, 16, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1]
[9, 16, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 16, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 17, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


### 10.2 Determine maximum architecture for 10 hidden layers

In [22]:
max_arch, max_edges = get_max_architecture(architectures_10)
print("For 10 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 10 hidden layers the architecture [9, 13, 5, 1, 1, 1, 1, 1, 1, 1, 1, 1] yields a maximum of 222 edges.


## 11 hidden layers

### 11.1 Generate all possible architectures for 11 hidden layers¶

In [23]:
def get_architectures_11_hidden_layers():
    my_list = []
    
    k = 11                  # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):
                            # for d6 only (n-d1-d2-d3-d4-d5) nodes are at our disposal
                            for d6 in range(1, n-d1-d2-d3-d4-d5):
                                # for d7 only (n-d1-d2-d3-d4-d5-d6) nodes are at our disposal
                                for d7 in range(1, n-d1-d2-d3-d4-d5-d6):
                                    # for d8 only (n-d1-d2-d3-d4-d5-d6-d7) nodes are at our disposal
                                    for d8 in range(1, n-d1-d2-d3-d4-d5-d6-d7):
                                        # for d9 only (n-d1-d2-d3-d4-d5-d6-d7-d8) nodes are at our disposal
                                        for d9 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8):
                                            # for d10 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9) nodes are at our disposal
                                            for d10 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9):

                                                d11 = n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10
                                                d = [9, d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, d11, 1]
                                                my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_11 = get_architectures_11_hidden_layers()
print(*architectures_11[:4], sep='\n')
print("...")
print(*architectures_11[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 15, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 14, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 13, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 12, 1]
...
[9, 14, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 14, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 14, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 15, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


### 11.2 Determine maximum architecture for 11 hidden layers

In [24]:
max_arch, max_edges = get_max_architecture(architectures_11)
print("For 11 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 11 hidden layers the architecture [9, 12, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] yields a maximum of 195 edges.


## 12 hidden layers

### 12.1 Generate all possible architectures for 12 hidden layers¶

In [25]:
def get_architectures_12_hidden_layers():
    my_list = []
    
    k = 12                  # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):
                            # for d6 only (n-d1-d2-d3-d4-d5) nodes are at our disposal
                            for d6 in range(1, n-d1-d2-d3-d4-d5):
                                # for d7 only (n-d1-d2-d3-d4-d5-d6) nodes are at our disposal
                                for d7 in range(1, n-d1-d2-d3-d4-d5-d6):
                                    # for d8 only (n-d1-d2-d3-d4-d5-d6-d7) nodes are at our disposal
                                    for d8 in range(1, n-d1-d2-d3-d4-d5-d6-d7):
                                        # for d9 only (n-d1-d2-d3-d4-d5-d6-d7-d8) nodes are at our disposal
                                        for d9 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8):
                                            # for d10 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9) nodes are at our disposal
                                            for d10 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9):
                                                # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9) nodes are at our disposal
                                                for d11 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10):

                                                    d12 = n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11
                                                    d = [9, d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, d11, d12, 1]
                                                    my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_12 = get_architectures_12_hidden_layers()
print(*architectures_12[:4], sep='\n')
print("...")
print(*architectures_12[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 13, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 12, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 11, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 10, 1]
...
[9, 12, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 12, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 12, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 13, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


### 12.2 Determine maximum architecture for 12 hidden layers

In [26]:
max_arch, max_edges = get_max_architecture(architectures_12)
print("For 12 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 12 hidden layers the architecture [9, 11, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] yields a maximum of 170 edges.


## 13 hidden layers

### 13.1 Generate all possible architectures for 13 hidden layers¶

In [27]:
def get_architectures_13_hidden_layers():
    my_list = []
    
    k = 13                  # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):
                            # for d6 only (n-d1-d2-d3-d4-d5) nodes are at our disposal
                            for d6 in range(1, n-d1-d2-d3-d4-d5):
                                # for d7 only (n-d1-d2-d3-d4-d5-d6) nodes are at our disposal
                                for d7 in range(1, n-d1-d2-d3-d4-d5-d6):
                                    # for d8 only (n-d1-d2-d3-d4-d5-d6-d7) nodes are at our disposal
                                    for d8 in range(1, n-d1-d2-d3-d4-d5-d6-d7):
                                        # for d9 only (n-d1-d2-d3-d4-d5-d6-d7-d8) nodes are at our disposal
                                        for d9 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8):
                                            # for d10 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9) nodes are at our disposal
                                            for d10 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9):
                                                # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9) nodes are at our disposal
                                                for d11 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10):
                                                    # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9) nodes are at our disposal
                                                    for d12 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11):
                                                    
                                                        d13 = n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12
                                                        d = [9, d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, d11, d12, d13, 1]
                                                        my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_13 = get_architectures_13_hidden_layers()
print(*architectures_13[:4], sep='\n')
print("...")
print(*architectures_13[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 11, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 10, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 9, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 8, 1]
...
[9, 10, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 10, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 10, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 11, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


### 13.2 Determine maximum architecture for 13 hidden layers

In [28]:
max_arch, max_edges = get_max_architecture(architectures_13)
print("For 13 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 13 hidden layers the architecture [9, 10, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] yields a maximum of 147 edges.


## 14 hidden layers

### 14.1 Generate all possible architectures for 14 hidden layers¶

In [29]:
def get_architectures_14_hidden_layers():
    my_list = []
    
    k = 14                  # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):
                            # for d6 only (n-d1-d2-d3-d4-d5) nodes are at our disposal
                            for d6 in range(1, n-d1-d2-d3-d4-d5):
                                # for d7 only (n-d1-d2-d3-d4-d5-d6) nodes are at our disposal
                                for d7 in range(1, n-d1-d2-d3-d4-d5-d6):
                                    # for d8 only (n-d1-d2-d3-d4-d5-d6-d7) nodes are at our disposal
                                    for d8 in range(1, n-d1-d2-d3-d4-d5-d6-d7):
                                        # for d9 only (n-d1-d2-d3-d4-d5-d6-d7-d8) nodes are at our disposal
                                        for d9 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8):
                                            # for d10 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9) nodes are at our disposal
                                            for d10 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9):
                                                # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10) nodes are at our disposal
                                                for d11 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10):
                                                    # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d9-d10-d11) nodes are at our disposal
                                                    for d12 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11):
                                                        # for d13 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12) nodes are at our disposal
                                                        for d13 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12):

                                                            d14 = n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13
                                                            d = [9, d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, d11, d12, d13, d14, 1]
                                                            my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_14 = get_architectures_14_hidden_layers()
print(*architectures_14[:4], sep='\n')
print("...")
print(*architectures_14[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 9, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 8, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 7, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 6, 1]
...
[9, 8, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 8, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 8, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


### 14.2 Determine maximum architecture for 14 hidden layers


In [30]:
max_arch, max_edges = get_max_architecture(architectures_14)
print("For 14 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 14 hidden layers the architecture [9, 9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] yields a maximum of 126 edges.


## 15 hidden layers

### 15.1 Generate all possible architectures for 15 hidden layers¶

In [31]:
def get_architectures_15_hidden_layers():
    my_list = []
    
    k = 15                  # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):
                            # for d6 only (n-d1-d2-d3-d4-d5) nodes are at our disposal
                            for d6 in range(1, n-d1-d2-d3-d4-d5):
                                # for d7 only (n-d1-d2-d3-d4-d5-d6) nodes are at our disposal
                                for d7 in range(1, n-d1-d2-d3-d4-d5-d6):
                                    # for d8 only (n-d1-d2-d3-d4-d5-d6-d7) nodes are at our disposal
                                    for d8 in range(1, n-d1-d2-d3-d4-d5-d6-d7):
                                        # for d9 only (n-d1-d2-d3-d4-d5-d6-d7-d8) nodes are at our disposal
                                        for d9 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8):
                                            # for d10 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9) nodes are at our disposal
                                            for d10 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9):
                                                # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10) nodes are at our disposal
                                                for d11 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10):
                                                    # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d9-d10-d11) nodes are at our disposal
                                                    for d12 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11):
                                                        # for d13 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12) nodes are at our disposal
                                                        for d13 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12):
                                                            # for d14 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13) nodes are at our disposal
                                                            for d14 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13):

                                                                d15 = n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14
                                                                d = [9, d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, d11, d12, d13, d14, d15, 1]
                                                                my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_15 = get_architectures_15_hidden_layers()
print(*architectures_15[:4], sep='\n')
print("...")
print(*architectures_15[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 7, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 6, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 5, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 4, 1]
...
[9, 6, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 6, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 6, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 7, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


### 15.2 Determine maximum architecture for 15 hidden layers

In [32]:
max_arch, max_edges = get_max_architecture(architectures_15)
print("For 15 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))

For 15 hidden layers the architecture [9, 7, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] yields a maximum of 106 edges.


## 16 hidden layers

### 16.1 Generate all possible architectures for 16 hidden layers¶

In [33]:
def get_architectures_16_hidden_layers():
    my_list = []
    
    k = 16                  # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):
                            # for d6 only (n-d1-d2-d3-d4-d5) nodes are at our disposal
                            for d6 in range(1, n-d1-d2-d3-d4-d5):
                                # for d7 only (n-d1-d2-d3-d4-d5-d6) nodes are at our disposal
                                for d7 in range(1, n-d1-d2-d3-d4-d5-d6):
                                    # for d8 only (n-d1-d2-d3-d4-d5-d6-d7) nodes are at our disposal
                                    for d8 in range(1, n-d1-d2-d3-d4-d5-d6-d7):
                                        # for d9 only (n-d1-d2-d3-d4-d5-d6-d7-d8) nodes are at our disposal
                                        for d9 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8):
                                            # for d10 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9) nodes are at our disposal
                                            for d10 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9):
                                                # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10) nodes are at our disposal
                                                for d11 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10):
                                                    # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d9-d10-d11) nodes are at our disposal
                                                    for d12 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11):
                                                        # for d13 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12) nodes are at our disposal
                                                        for d13 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12):
                                                            # for d14 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13) nodes are at our disposal
                                                            for d14 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13):
                                                                # for d15 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14) nodes are at our disposal
                                                                for d15 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14):

                                                                    d16 = n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14-d15
                                                                    d = [9, d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, d11, d12, d13, d14, d15, d16, 1]
                                                                    my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_16 = get_architectures_16_hidden_layers()
print(*architectures_16[:4], sep='\n')
print("...")
print(*architectures_16[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 5, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 4, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 3, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 2, 1]
...
[9, 4, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 4, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 4, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 5, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


### 16.2 Determine maximum architecture for 16 hidden layers


In [34]:
max_arch, max_edges = get_max_architecture(architectures_16)
print("For 16 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))


For 16 hidden layers the architecture [9, 5, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] yields a maximum of 86 edges.


## 17 hidden layers

### 17.1 Generate all possible architectures for 17 hidden layers¶

In [35]:
def get_architectures_17_hidden_layers():
    my_list = []
    
    k = 17                  # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):
                            # for d6 only (n-d1-d2-d3-d4-d5) nodes are at our disposal
                            for d6 in range(1, n-d1-d2-d3-d4-d5):
                                # for d7 only (n-d1-d2-d3-d4-d5-d6) nodes are at our disposal
                                for d7 in range(1, n-d1-d2-d3-d4-d5-d6):
                                    # for d8 only (n-d1-d2-d3-d4-d5-d6-d7) nodes are at our disposal
                                    for d8 in range(1, n-d1-d2-d3-d4-d5-d6-d7):
                                        # for d9 only (n-d1-d2-d3-d4-d5-d6-d7-d8) nodes are at our disposal
                                        for d9 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8):
                                            # for d10 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9) nodes are at our disposal
                                            for d10 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9):
                                                # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10) nodes are at our disposal
                                                for d11 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10):
                                                    # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d9-d10-d11) nodes are at our disposal
                                                    for d12 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11):
                                                        # for d13 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12) nodes are at our disposal
                                                        for d13 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12):
                                                            # for d14 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13) nodes are at our disposal
                                                            for d14 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13):
                                                                # for d15 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14) nodes are at our disposal
                                                                for d15 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14):
                                                                    # for d16 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14-d15) nodes are at our disposal
                                                                    for d16 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14-d15):

                                                                        d17 = n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14-d15-d16
                                                                        d = [9, d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, d11, d12, d13, d14, d15, d16, d17, 1]
                                                                        my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_17 = get_architectures_17_hidden_layers()
print(*architectures_17[:4], sep='\n')
print("...")
print(*architectures_17[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1]
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 2, 1]
...
[9, 2, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 2, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
[9, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


### 17.2 Determine maximum architecture for 17 hidden layers



In [36]:
max_arch, max_edges = get_max_architecture(architectures_17)
print("For 17 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))


For 17 hidden layers the architecture [9, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] yields a maximum of 66 edges.


## 18 hidden layers

### 18.1 Generate all possible architectures for 18 hidden layers¶

In [37]:
def get_architectures_18_hidden_layers():
    my_list = []
    
    k = 18                  # k hidden layers
    n = 36-k                # number of regular nodes, 
                            # i.e. total units minus k bias nodes    
    
    for d1 in range(1, n):
        # for d2 only (n-d1) nodes are at our disposal
        for d2 in range(1, n-d1):
            # for d3 only (n-d1-d2) nodes are at our disposal
            for d3 in range(1, n-d1-d2):
                # for d4 only (n-d1-d2-d3) nodes are at our disposal
                for d4 in range(1, n-d1-d2-d3):
                    # for d5 only (n-d1-d2-d3-d4) nodes are at our disposal
                        for d5 in range(1, n-d1-d2-d3-d4):
                            # for d6 only (n-d1-d2-d3-d4-d5) nodes are at our disposal
                            for d6 in range(1, n-d1-d2-d3-d4-d5):
                                # for d7 only (n-d1-d2-d3-d4-d5-d6) nodes are at our disposal
                                for d7 in range(1, n-d1-d2-d3-d4-d5-d6):
                                    # for d8 only (n-d1-d2-d3-d4-d5-d6-d7) nodes are at our disposal
                                    for d8 in range(1, n-d1-d2-d3-d4-d5-d6-d7):
                                        # for d9 only (n-d1-d2-d3-d4-d5-d6-d7-d8) nodes are at our disposal
                                        for d9 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8):
                                            # for d10 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9) nodes are at our disposal
                                            for d10 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9):
                                                # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10) nodes are at our disposal
                                                for d11 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10):
                                                    # for d12 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d9-d10-d11) nodes are at our disposal
                                                    for d12 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11):
                                                        # for d13 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12) nodes are at our disposal
                                                        for d13 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12):
                                                            # for d14 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13) nodes are at our disposal
                                                            for d14 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13):
                                                                # for d15 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14) nodes are at our disposal
                                                                for d15 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14):
                                                                    # for d16 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14-d15) nodes are at our disposal
                                                                    for d16 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14-d15):
                                                                        # for d17 only (n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14-d15-d16) nodes are at our disposal
                                                                        for d17 in range(1, n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14-d15-d16):
                                                                        
                                                                            d18 = n-d1-d2-d3-d4-d5-d6-d7-d8-d9-d10-d11-d12-d13-d14-d15-d16-d17
                                                                            d = [9, d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, d11, d12, d13, d14, d15, d16, d17, d18, 1]
                                                                            my_list.append(d)

    # check if we considered all possible compositions
    assert(num_compositions(n,k) == len(my_list))   
    return my_list

#------------------

architectures_18 = get_architectures_18_hidden_layers()
print(*architectures_18[:4], sep='\n')
print("...")
print(*architectures_18[-4:], sep='\n')

[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
...
[9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


### 18.2 Determine maximum architecture for 18 hidden layers



In [38]:
max_arch, max_edges = get_max_architecture(architectures_18)
print("For 18 hidden layers the architecture {0} yields a maximum of {1} edges.".format(max_arch, max_edges))


For 18 hidden layers the architecture [9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] yields a maximum of 46 edges.


# Conclusion

- Using a brute force approach we determined that a minimum of 46 edges is achieved for 18 hidden layers with the architecture $d = [9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]$.

- A maximum of 510 edges is achieved for 2 hidden layers with the the architecture $d = [9, 21, 13, 1]$.